### PR TITLE
fix(speculative): embed d2t/t2d vocab remap in EAGLE-3 draft checkpoint

### PR DESCRIPTION
**Pull request by khazic**
## What

Embed the EAGLE-3 draft-to-target vocab remap tables (`d2t` / `t2d`) into the saved draft checkpoint so inference engines can map compressed-draft logits back to the target vocab.

## Why

When the draft vocab is compressed (`draft_vocab_size < target vocab_size`, the frequency-pruned EAGLE-3 setup), the draft `lm_head` emits logits over the **pruned** draft vocab. To verify draft tokens against the target, the inference engine must remap each draft id back to the target vocab:

- vLLM loads `d2t` (renamed to `draft_id_to_target_id`) and computes `target_id = draft_id + d2t[draft_id]`.
- SGLang uses the boolean `t2d` mask over the full target vocab.

Previously Automodel saved only `selected_token_ids` into the side file `eagle_meta.pt`, never into the model weights. vLLM therefore found no mapping in `model.safetensors`, silently aligned draft ids to the first `draft_vocab_size` target ids, and the acceptance rate collapsed at inference. (SGLang served through `serve_sglang.py` was unaffected only because that path injects a separate `--speculative-token-map`; native SGLang loading hit the same problem.)

## What changed

- `LlamaEagle3DraftModel` now registers persistent `d2t` / `t2d` buffers, **only** when the vocab is compressed (with `draft_vocab_size == vocab_size` vLLM expects no mapping). A new `set_vocab_mapping(selected_token_ids)` populates them: `d2t[i] = selected_token_ids[i] - i` and `t2d[selected_token_ids] = True`.
- The EAGLE-3 recipe calls `set_vocab_mapping` right after building the draft, so the saved `model.safetensors` is self-contained for vLLM and native SGLang.
- The `GptOss` draft subclass inherits the buffers, and both the TTT and P-EAGLE paths save through the same draft model, so both are covered.

## Scope / follow-ups

- This fixes **newly trained** checkpoints. Existing checkpoints that only carry `eagle_meta.pt` still need a one-off conversion to inject `d2t` / `t2d` into their safetensors.
- `serve_sglang.py` still regenerates a token map from `eagle_meta.pt`; it can later read `t2d` straight from the safetensors and retire that workaround. Left out of this PR to keep it focused.

## Tests

Added unit tests in `tests/unit_tests/speculative/test_eagle3.py`:

- buffers are registered with correct shape/dtype and land in `state_dict` only when compressed;
- `set_vocab_mapping` builds the correct offset (`d2t`) and presence mask (`t2d`);
- wrong-length input raises;
- integer/bool buffers survive a `.to(dtype=...)` cast;
- buffers round-trip exactly through `state_dict`.

`tests/unit_tests/speculative/` all green (294 passed), ruff clean.

### CODE REVIEW

File: tests/unit_tests/speculative/test_eagle3.py [hunk 1]
```diff
@@ -1205,3 +1209,76 @@ def test_eagle3_1_trainer_multi_step_runs_with_flags_enabled():
...
+    keys = model.state_dict().keys()
+    assert "d2t" not in keys and "t2d" not in keys
+    # No-op even when called, so callers need no compression branch of their own.
+    model.set_vocab_mapping(torch.arange(64, dtype=torch.long))
+
+
+def test_set_vocab_mapping_builds_offset_and_mask():
+    """d2t is the draft->target offset; t2d marks selected target ids."""
+    model = LlamaEagle3DraftModel(_make_draft_config(vocab_size=128, draft_vocab_size=4))
+    selected = torch.tensor([5, 9, 9, 100], dtype=torch.long)  # draft id -> target id
```
[the remark this answers is not in the document]
Comment (khazic) [reply tests/unit_tests/speculative/test_eagle3.py]: Good catch, fixed in fee29340. Switched to [5, 9, 50, 100] so every target slot is distinct and t2d now has 4 True entries, which exercises the off-by-one you mentioned.

**Review by HuiyingLi — APPROVED** [review]
